test/lightning/workflow_versions_test.exs

```diff
@@ -46,10 +46,115 @@ defmodule Lightning.WorkflowVersionsTest do
       assert {:error, :invalid_input} =
                WorkflowVersions.record_version(wf, @a, "web")
     end
+
+    test "does not insert duplicate when hash is same as latest" do
+      wf = insert(:workflow)
+
+      # Insert first version
+      assert {:ok, wf1} = WorkflowVersions.record_version(wf, @a, "app")
+      assert wf1.version_history == [@a]
+      assert count_rows(wf.id) == 1
+
+      # Insert different hash
+      assert {:ok, wf2} = WorkflowVersions.record_version(wf1, @b, "cli")
+      assert wf2.version_history == [@a, @b]
+      assert count_rows(wf.id) == 2
+
+      # Try to insert the same hash as latest (duplicate)
+      assert {:ok, wf3} = WorkflowVersions.record_version(wf2, @b, "app")
+      assert wf3.version_history == [@a, @b]
+      # Should still be 2, no new row inserted
+      assert count_rows(wf.id) == 2
+
+      # Try with different source but same hash - still a duplicate
+      assert {:ok, wf4} = WorkflowVersions.record_version(wf3, @b, "cli")
+      assert wf4.version_history == [@a, @b]
+      # Still no new row
+      assert count_rows(wf.id) == 2
+    end
+
+    test "squashes when source is same as latest version's source" do
+      wf = insert(:workflow)
+
+      # Insert first version from app
+      assert {:ok, wf1} = WorkflowVersions.record_version(wf, @a, "app")
+      assert wf1.version_history == [@a]
+      assert count_rows(wf.id) == 1
+
+      # Insert second version from cli
+      assert {:ok, wf2} = WorkflowVersions.record_version(wf1, @b, "cli")
+      assert wf2.version_history == [@a, @b]
+      assert count_rows(wf.id) == 2
+
+      # Insert third version from cli (same source as latest) - should squash
+      assert {:ok, wf3} = WorkflowVersions.record_version(wf2, @c, "cli")
+      # @b should be replaced by @c
```

Comment:
This way of articulating the tests is absolutely brilliant - thank you! :trophy: